src/test/java/com/google/sps/data/DatabaseHandlerTest.java

```diff
@@ -0,0 +1,97 @@
+public class DatabaseHandlerTest {
+  private final LocalServiceTestHelper helper = new LocalServiceTestHelper(new LocalDatastoreServiceTestConfig());
+  private DatabaseHandler databaseHandler = new DatabaseHandler();
+
+  @Before
+  public void setUp() {
+    helper.setUp();
+  }
+
+  @Test
+  public void testAddingBookToDatastore() {
+    databaseHandler.addBookStockToDatastore("12345");
+    Query query = new Query("Book").setFilter(new Query.FilterPredicate("bookId", Query.FilterOperator.EQUAL, "12345"));
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+    Entity bookEntity = results.asSingleEntity();
+    Long expectedTotalStock = (long) 185;
+    // Check if total stock was added correctly to Datastore.
+    Assert.assertEquals(expectedTotalStock, bookEntity.getProperty("totalStock"));
+    Query librariesQuery = new Query("LibraryStock").setAncestor(bookEntity.getKey());
+    int expectedNumberOfLibraries = 37;
+    // Check if all children Libraries were added succesfully.
+    Assert.assertEquals(expectedNumberOfLibraries,
+        datastore.prepare(librariesQuery).asList(FetchOptions.Builder.withDefaults()).size());
+    Query individualLibrariyQuery = new Query("LibraryStock").setAncestor(bookEntity.getKey())
+        .setFilter(new Query.FilterPredicate("libraryId", Query.FilterOperator.EQUAL, 1));
+    Long expectedNumberOfBooks = (long) 5;
+    // Check if individual libraries have the correct number of books.
+    Assert.assertEquals(expectedNumberOfBooks,
+        datastore.prepare(individualLibrariyQuery).asSingleEntity().getProperty("stock"));
+  }
+
+  @Test
+  public void testLoadingDataFromCSV() throws URISyntaxException, IOException {
+    File csvFile = new File(this.getClass().getResource("/examples.csv").toURI());
+    databaseHandler.loadBookStocksFromCSV(csvFile);
+    Query query = new Query("Book").setFilter(new Query.FilterPredicate("bookId", Query.FilterOperator.EQUAL, "gqX7rQEACAAJ"));
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    // Check if book was added to datastore.
+    Assert.assertEquals(1, datastore.prepare(query).asList(FetchOptions.Builder.withDefaults()).size());
+  }
+
+  @Test
+  public void testLoadingDuplicateDataFromCSV() throws URISyntaxException, IOException {
+    File csvFile = new File(this.getClass().getResource("/examples.csv").toURI());
+    // Load CSV twice
+    databaseHandler.loadBookStocksFromCSV(csvFile);
+    databaseHandler.loadBookStocksFromCSV(csvFile);
+    Query query = new Query("Book").setFilter(new Query.FilterPredicate("bookId", Query.FilterOperator.EQUAL, "gqX7rQEACAAJ"));
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    // Check if book was added just once to datastore.
+    Assert.assertEquals(1, datastore.prepare(query).asList(FetchOptions.Builder.withDefaults()).size());
+  }
+
+    
+  @After
```

Comment:
nit: generally you'd put this method right affer setup()

This actually runs after each individual test, so it's presence at the bottom of the file isn't really more intuitive.

Reply:
Fixed